ext/mbstring/libmbfl/filters/mbfilter_utf7imap.c

```diff
@@ -418,3 +426,175 @@ int mbfl_filt_conv_wchar_utf7imap_flush(mbfl_convert_filter *filter)
+	} \
+	filter->status = (cache << (8 + shift1)) | (EXTRACT_BITS(bits, shift1) << 8) | (state + offset); \
+	break;
+
+	int cache = filter->status >> 8, state = filter->status & 0xFF;
+	switch (state) {
+		/* 1st 16 bits of a UTF-16 code unit, might be surrogate pair */
+		case 1: /* 0 bits consumed already */
+			CHECK_SURROGATE_FIRST_PART(six_bits, 0, 0, 3, 10);
+		case 2: /* 2 bits consumed already */
+			CHECK_SURROGATE_FIRST_PART(six_bits, 2, 0, 3, 10);
+		case 3: /* 4 bits... */
+			CHECK_SURROGATE_FIRST_PART(six_bits, 4, 0, 3, 10);
+		case 4: /* 6 bits... */
+			/* We have definitely found a surrogate pair
+			 * Remove the top 6 cached bits which have the 'tag' for a surrogate pair */
+			filter->status = (six_bits << 8) | (state + 3); break;
+		case 5: /* 8 bits... */
+			/* Likewise */
+			filter->status = ((cache & 0x3) << 14) | (six_bits << 8) | (state + 3); break;
+		case 6: /* 10 bits... */
+			filter->status = ((cache & 0xF) << 14) | (six_bits << 8) | (state + 10); break;
+		case 7: /* 12 bits... */
+			CHECK_SURROGATE_SECOND_PART(six_bits, 0, 4, 10);
+		case 8: /* 14 bits... */
+			CHECK_SURROGATE_SECOND_PART(six_bits, 0, 2, 10);
+
+		/* 1st 16 bits of a UTF-16 code unit; definitely not a surrogate pair */
+		case 9: /* 2 bits consumed already */
+		case 10: /* 4 bits... */
+		case 11: /* 6 bits... */
+		case 12: /* 8 bits... */
+			filter->status = (cache << 14) | (six_bits << 8) | (state + 3); break;
+		case 13: /* 10 bits... */
+			check_legal_codepoint_for_base64((cache << 6) | six_bits, filter);
+			/* We will end exactly at a character boundary, so there is no need to keep cached bits */
+			filter->status = 1; break;
+		case 14: /* 12 bits... */
+			check_legal_codepoint_for_base64((cache << 4) | (six_bits >> 2), filter);
+			CHECK_SURROGATE_FIRST_PART(six_bits, 0, 4, -12, -5);
+		case 15: /* 14 bits... */
+			check_legal_codepoint_for_base64((cache << 2) | (six_bits >> 4), filter);
+			CHECK_SURROGATE_FIRST_PART(six_bits, 0, 2, -12, -5);
+
+		/* 2nd 16 bits of a surrogate pair */
+		case 16: /* 0 bits consumed already */
+			CHECK_SURROGATE_SECOND_PART(six_bits, 0, 0, 3);
+		case 17: /* 2 bits... */
+			CHECK_SURROGATE_SECOND_PART(six_bits, 2, 0, 3);
+		case 18: /* 4 bits... */
+			CHECK_SURROGATE_SECOND_PART(six_bits, 4, 0, 3);
+		case 19: /* 6 bits... */
+		case 20: /* 8 bits... */
+			filter->status = (cache << 14) | (six_bits << 8) | (state + 3); break;
+		case 21: /* 10 bits... */
+			check_legal_codepoint_for_base64((cache << 6) | six_bits, filter);
+			/* We will end exactly at a character boundary, so there is no need to keep cached bits */
+			filter->status = 1; break;
+		case 22: /* 12 bits... */
+			check_legal_codepoint_for_base64((cache << 4) | (six_bits >> 2), filter);
+			CHECK_SURROGATE_FIRST_PART(six_bits, 0, 4, -20, -13);
+		case 23: /* 14 bits... */
+			check_legal_codepoint_for_base64((cache << 2) | (six_bits >> 4), filter);
+			CHECK_SURROGATE_FIRST_PART(six_bits, 0, 2, -20, -13);
+	}
+	return c;
+}
```

Comment:
I'm concerned about this one. As mentioned before, it should be possible to implement identification based on the input filter, with some loss of efficiency. For simple encodings, reimplementing the logic as an identify filter is not a big deal, but for non-trivial encodings like this one, this results in significant duplication of logic that doesn't serve a lot of purpose (I don't think high-performance identification of utf-7 is a common use-case...)

Reply:
> ...it should be possible to implement identification based on the input filter...

For UTF-7 (especially UTF7-IMAP), that is not the case.

The conversion filters do keep a count of "illegal characters", which can be consulted after an operation finishes. If an "illegal character" is encountered, they also can take other action, such as inserting a placeholder (such as `?`) into the output.

However, not all invalid strings have illegal characters in them! Some encodings have multiple "modes", with escape sequences to switch from one mode to another. The spec for UTF7-IMAP explicitly states that valid strings must end in ASCII mode. So a UTF7-IMAP string can be invalid, without containing any invalid characters. Further, the spec for UTF7-IMAP explicitly bans strings which shift into Base64 mode and immediately back to ASCII mode without any intervening Base64-encoded characters. And in addition, it is also illegal to shift from Base64 mode to ASCII mode and immediately back to Base64 mode without any intervening ASCII characters. If you're not bored yet, let me mention that it is also illegal to Base64-encode characters which *could* have been represented in ASCII. (The point of all these restrictions is that UTF7-IMAP has one, and only one, valid representation for any sequence of characters.)

Likewise, a string in ISO-2022 (or similar) encoding could contain an unrecognized escape sequence. Do we count that as an "illegal character", and potentially emit `?` to the output? I think not.

However, you are very right that for *some* encodings, the conversion and identify filters could be unified. This will require further refactoring, though.

Reply:
@nikic, let me say another thing: If you feel that my code for verifying UTF7-IMAP is too hairy and can be done better, please say so. I'm not saying that the code can't be improved; but I think trying to combine the identify and conversion filters for this encoding is a non-starter.

Reply:
> > ...it should be possible to implement identification based on the input filter...
> 
> For UTF-7 (especially UTF7-IMAP), that is not the case.
> 
> The conversion filters do keep a count of "illegal characters", which can be consulted after an operation finishes. If an "illegal character" is encountered, they also can take other action, such as inserting a placeholder (such as `?`) into the output.
> 
> However, not all invalid strings have illegal characters in them! Some encodings have multiple "modes", with escape sequences to switch from one mode to another. The spec for UTF7-IMAP explicitly states that valid strings must end in ASCII mode. So a UTF7-IMAP string can be invalid, without containing any invalid characters. Further, the spec for UTF7-IMAP explicitly bans strings which shift into Base64 mode and immediately back to ASCII mode without any intervening Base64-encoded characters. And in addition, it is also illegal to shift from Base64 mode to ASCII mode and immediately back to Base64 mode without any intervening ASCII characters. If you're not bored yet, let me mention that it is also illegal to Base64-encode characters which _could_ have been represented in ASCII. (The point of all these restrictions is that UTF7-IMAP has one, and only one, valid representation for any sequence of characters.)
> 
> Likewise, a string in ISO-2022 (or similar) encoding could contain an unrecognized escape sequence. Do we count that as an "illegal character", and potentially emit `?` to the output? I think not.
> 
> However, you are very right that for _some_ encodings, the conversion and identify filters could be unified. This will require further refactoring, though.

Interesting case... I still think that the conversion filter needs to have the ability to recognize such "state" errors as well. Keeping in mind that the implementation of mb_check_encoding() is actually base on the conversion filter, *not* the identify filter, the conversion filter needs to recognize this to produce a correct mb_check_encoding() result.

Generally, I believe there must be an equivalence relation here, in that the conversion filter must report an error if and only if the identify filter reports an error, and vice versa. Otherwise there will be discrepancies, like a string being recognized by mb_detect_encoding in strict mode, but not passing mb_check_encoding, or the other way around.

Now, that still leaves the question of what constitutes an "error". We could handle these state errors by emitting a substitute character at the point where the error is detected, but that would be distinctly weird. I think it would be better to simply increment the number of illegal characters without emitting a substitution character. Alternatively we could add an additional error count for non-character errors, but I'm not sure the distinction would be useful for anything.